File: Web/DAL/Factory/VoteFactory.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Web;
using Web.Models;

namespace Web.DAL.Factory
{
    public interface IVoteType
    {
        int Limit { get; }
        void SaveData();
        void Apply();
    }

    public enum VoteType
    {
        FirstPass,
        Preferential,
        Transferable
    }

    public class VoteFactory
    {
        public IVoteType Create(VoteType voteType)
        {
            switch (voteType)
            {
                case VoteType.Preferential:
                    return new Preferential();
                case VoteType.Transferable:
                    return new Transferable();
                case VoteType.FirstPass:
                    return new FirstPast();
                default:
                    throw new NotImplementedException();
            }
        }
    }
}

File: server/src/factory/VoteFactory.ts
// superherofactory.ts
import { FirstPast } from '../firstpass/firstpass.model';
import { Transferable } from '../transferable/transferable.model';
import { Preferential } from '../preferential/preferential.model';
import { NotImplementedException } from '@nestjs/common';


export interface IVoteType
    {
        startDate : Date;
        EndDate : Date;
        Limit : number; 
    }


export enum VoteTypes {
    FirstPast,
    Transferable,
    Preferential
}

export class VoteTypeFactory implements IVoteType {

    startDate : Date;
    EndDate : Date;
    Limit : number; 

    create(VoteTypes);
    create(VoteTypes): FirstPast;
    create(VoteTypes): Transferable;
    create(VoteTypes): Preferential;


  public create(VoteTypes): FirstPast | Transferable | Preferential {
    switch(VoteTypes)
    {
        case VoteTypes.FirstPast:
            return new FirstPast();
        case VoteTypes.Transferable:
            return new Transferable();
        case VoteTypes.Preferential:
            return new Preferential();
        default:
        throw new NotImplementedException();
            
    }
  }
}